File: sdks/csharp/Goodbase/GoodbaseTelemetry.cs
using System.Diagnostics;
using System.Text.Json;

namespace Goodbase;

public enum GoodbaseConsent { Granted, Essential, Denied }

public sealed record GoodbaseTelemetryOptions(string AppId, string Release, string BuildNumber, string StorageDirectory, string? DistributionTrack = null);

public sealed class GoodbaseTelemetry : IDisposable
{
    private readonly GoodbaseClient _client;
    private readonly GoodbaseTelemetryOptions _options;
    private readonly string _sessionId = Guid.NewGuid().ToString();
    private readonly string _spool;
    private readonly object _sync = new();
    private readonly List<Dictionary<string, object?>> _breadcrumbs = [];
    private readonly Dictionary<string, string> _customKeys = [];
    private bool _started;
    public GoodbaseConsent Consent { get; private set; }

    public GoodbaseTelemetry(GoodbaseClient client, GoodbaseTelemetryOptions options, GoodbaseConsent consent = GoodbaseConsent.Denied)
    {
        _client=client; _options=options; Consent=consent;
        Directory.CreateDirectory(options.StorageDirectory);
        _spool=Path.Combine(options.StorageDirectory, $"goodbase-{Sanitize(options.AppId)}.jsonl");
    }

    public void Start()
    {
        if (_started || Consent == GoodbaseConsent.Denied) return;
        _started=true; AppDomain.CurrentDomain.UnhandledException += OnUnhandledException;
        TaskScheduler.UnobservedTaskException += OnUnobservedTaskException;
        Send("session", SessionPayload("start")); _ = FlushAsync();
    }

    public void SetConsent(GoodbaseConsent value)
    {
        Consent=value;
        if (value == GoodbaseConsent.Denied) { if(File.Exists(_spool))File.Delete(_spool); Dispose(); } else Start();
    }

    public void Breadcrumb(string message, IReadOnlyDictionary<string, object?>? data = null)
    {
        lock(_sync){_breadcrumbs.Add(new(){["message"]=message[..Math.Min(message.Length,500)],["data"]=data,["at"]=DateTimeOffset.UtcNow});if(_breadcrumbs.Count>50)_breadcrumbs.RemoveAt(0);}
    }

    public void SetCustomKey(string key, object? value) { lock(_sync){if(_customKeys.ContainsKey(key)||_customKeys.Count<64){var text=value?.ToString()??"";_customKeys[key[..Math.Min(key.Length,100)]]=text[..Math.Min(text.Length,1000)];}} }
    public void CaptureException(Exception error, bool fatal=false, string? type=null) => Send("crash", CrashPayload(error,fatal,type));

    public async Task<T> TraceAsync<T>(string name, Func<Task<T>> operation, string type="custom")
    {
        var watch=Stopwatch.StartNew();try{return await operation();}catch(Exception error){CaptureException(error);throw;}finally{Send("trace",new{appId=_options.AppId,type,name,durationMs=watch.Elapsed.TotalMilliseconds,occurredAt=DateTimeOffset.UtcNow});}
    }

    public async Task FlushAsync(CancellationToken cancellationToken=default)
    {
        if(Consent==GoodbaseConsent.Denied||!File.Exists(_spool))return;
        List<JsonElement> remaining=[];
        string[] lines;lock(_sync)lines=File.ReadAllLines(_spool);
        foreach(var line in lines){try{using var document=JsonDocument.Parse(line);await UploadAsync(document.RootElement,cancellationToken);}catch{try{using var document=JsonDocument.Parse(line);remaining.Add(document.RootElement.Clone());}catch{}}}
        Replace(remaining.Select(value => JsonSerializer.Serialize(value)).TakeLast(100));
    }

    private object SessionPayload(string action,string? endedReason=null)=>new{appId=_options.AppId,sessionId=_sessionId,action,consentState=Consent.ToString().ToLowerInvariant(),occurredAt=DateTimeOffset.UtcNow,release=_options.Release,buildNumber=_options.BuildNumber,distributionTrack=_options.DistributionTrack,endedReason};
    private object CrashPayload(Exception error,bool fatal,string? type){lock(_sync)return new{appId=_options.AppId,platform="dotnet",occurredAt=DateTimeOffset.UtcNow,title=error.Message,stackTrace=error.ToString(),sessionId=_sessionId,release=_options.Release,buildNumber=_options.BuildNumber,fatal,exceptionType=type??error.GetType().FullName,breadcrumbs=_breadcrumbs.ToArray(),customKeys=new Dictionary<string,string>(_customKeys),device=new{os=Environment.OSVersion.ToString(),runtime=Environment.Version.ToString()}};}
    private void Send(string kind,object payload){if(Consent==GoodbaseConsent.Denied)return;var json=JsonSerializer.Serialize(new{kind,payload});_ = Task.Run(async()=>{try{using var document=JsonDocument.Parse(json);await UploadAsync(document.RootElement);}catch{Append(json);}});}
    private async Task UploadAsync(JsonElement eventValue,CancellationToken cancellationToken=default){var kind=eventValue.GetProperty("kind").GetString();var body=eventValue.GetProperty("payload").GetRawText();if(kind=="session")await _client.RecordSessionAsync(body,cancellationToken);else if(kind=="crash")await _client.CaptureCrashAsync(body,cancellationToken);else await _client.RecordTraceAsync(body,cancellationToken);}
    private void Append(string line){lock(_sync){var lines=File.Exists(_spool)?File.ReadAllLines(_spool).ToList():[];lines.Add(line);Replace(lines.TakeLast(100));}}
    private void Replace(IEnumerable<string> lines){var temporary=_spool+".tmp";File.WriteAllLines(temporary,lines);File.Move(temporary,_spool,true);}
    private void OnUnhandledException(object sender,UnhandledExceptionEventArgs args){if(args.ExceptionObject is Exception error){var json=JsonSerializer.Serialize(new{kind="crash",payload=CrashPayload(error,true,"UnhandledException")});Append(json);}}
    private void OnUnobservedTaskException(object? sender,UnobservedTaskExceptionEventArgs args){CaptureException(args.Exception,false,"UnobservedTaskException");}
    private static string Sanitize(string value)=>string.Concat(value.Select(ch=>char.IsLetterOrDigit(ch)||"._-".Contains(ch)?ch:'_'));
    public void Dispose(){if(!_started)return;Send("session",SessionPayload("end","normal"));AppDomain.CurrentDomain.UnhandledException-=OnUnhandledException;TaskScheduler.UnobservedTaskException-=OnUnobservedTaskException;_started=false;}
}

File: sdks/dart/lib/goodbase_telemetry.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:ui';
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'goodbase.dart';

enum GoodbaseConsent { granted, essential, denied }

class GoodbaseTelemetryConfig {
  const GoodbaseTelemetryConfig({required this.appId, required this.release, required this.buildNumber, this.distributionTrack, required this.storageDirectory});
  final String appId;
  final String release;
  final String buildNumber;
  final String? distributionTrack;
  final Directory storageDirectory;
}

class GoodbaseTelemetry with WidgetsBindingObserver {
  GoodbaseTelemetry(this.client, this.config, {this.consent = GoodbaseConsent.denied})
      : sessionId = '${DateTime.now().microsecondsSinceEpoch}-${Platform.localHostname}',
        _buffer = File('${config.storageDirectory.path}/goodbase-${config.appId.replaceAll(RegExp(r'[^A-Za-z0-9._-]'), '_')}.jsonl');

  final GoodbaseClient client;
  final GoodbaseTelemetryConfig config;
  final String sessionId;
  GoodbaseConsent consent;
  final File _buffer;
  final List<Map<String, Object?>> _breadcrumbs = [];
  final Map<String, String> _customKeys = {};
  FlutterExceptionHandler? _previousFlutterHandler;
  ErrorCallback? _previousPlatformHandler;
  Timer? _flushTimer;
  bool _started = false;

  Future<void> start() async {
    if (_started || consent == GoodbaseConsent.denied) return;
    _started = true;
    WidgetsBinding.instance.addObserver(this);
    _previousFlutterHandler = FlutterError.onError;
    FlutterError.onError = (details) { capture(details.exception, details.stack, fatal: true, type: 'FlutterError'); _previousFlutterHandler?.call(details); };
    _previousPlatformHandler = PlatformDispatcher.instance.onError;
    PlatformDispatcher.instance.onError = (error, stack) { capture(error, stack, fatal: true, type: 'PlatformError'); return _previousPlatformHandler?.call(error, stack) ?? false; };
    await _session('start');
    await flush();
    _flushTimer = Timer.periodic(const Duration(seconds: 30), (_) => flush());
  }

  Future<void> stop() async {
    if (!_started) return;
    await _session('end', endedReason: 'normal');
    WidgetsBinding.instance.removeObserver(this);
    FlutterError.onError = _previousFlutterHandler;
    PlatformDispatcher.instance.onError = _previousPlatformHandler;
    _flushTimer?.cancel();
    _started = false;
  }

  Future<void> setConsent(GoodbaseConsent value) async {
    consent = value;
    if (value == GoodbaseConsent.denied) { if (await _buffer.exists()) await _buffer.delete(); await stop(); } else { await start(); }
  }

  void breadcrumb(String message, [Map<String, Object?> data = const {}]) {
    _breadcrumbs.add({'message': message.length > 500 ? message.substring(0, 500) : message, 'data': data, 'at': DateTime.now().toUtc().toIso8601String()});
    if (_breadcrumbs.length > 50) _breadcrumbs.removeAt(0);
  }

  void setCustomKey(String key, Object? value) { if (_customKeys.containsKey(key) || _customKeys.length < 64) _customKeys[key] = value.toString(); }

  Future<void> capture(Object error, StackTrace? stack, {bool fatal = false, String? type}) => _send('crash', {
    'appId': config.appId, 'platform': 'flutter', 'occurredAt': DateTime.now().toUtc().toIso8601String(),
    'title': error.toString(), 'stackTrace': stack?.toString(), 'sessionId': sessionId, 'release': config.release,
    'buildNumber': config.buildNumber, 'fatal': fatal, 'exceptionType': type ?? error.runtimeType.toString(),
    'breadcrumbs': _breadcrumbs, 'customKeys': _customKeys,
    'device': {'os': Platform.operatingSystem, 'osVersion': Platform.operatingSystemVersion, 'locale': PlatformDispatcher.instance.locale.toLanguageTag()}
  });

  Future<T> trace<T>(String name, Future<T> Function() operation, {String type = 'custom'}) async {
    final watch = Stopwatch()..start();
    try { return await operation(); } catch (error, stack) { await capture(error, stack, type: 'nonfatal'); rethrow; }
    finally { await _send('trace', {'appId': config.appId, 'type': type, 'name': name, 'durationMs': watch.elapsedMicroseconds / 1000, 'occurredAt': DateTime.now().toUtc().toIso8601String()}); }
  }

  Future<void> flush() async {
    if (consent == GoodbaseConsent.denied || !await _buffer.exists()) return;
    final events = (await _buffer.readAsLines()).map((line) { try { return jsonDecode(line) as Map<String, dynamic>; } catch (_) { return null; } }).whereType<Map<String, dynamic>>();
    final remaining = <Map<String, dynamic>>[];
    for (final event in events) { try { await _upload(event); } catch (_) { remaining.add(event); } }
    await _replace(remaining.take(100).toList());
  }

  Future<void> _session(String action, {String? endedReason}) => _send('session', {'appId': config.appId, 'sessionId': sessionId, 'action': action, 'consentState': consent.name, 'occurredAt': DateTime.now().toUtc().toIso8601String(), 'release': config.release, 'buildNumber': config.buildNumber, 'distributionTrack': config.distributionTrack, 'endedReason': endedReason});
  Future<void> _send(String kind, Map<String, Object?> payload) async { if (consent == GoodbaseConsent.denied) return; final event={'kind':kind,'payload':payload}; try { await _upload(event); } catch (_) { await _append(event); } }
  Future<void> _upload(Map<String, dynamic> event) { final payload=Map<String,dynamic>.from(event['payload'] as Map); return event['kind']=='crash' ? client.captureCrash(payload) : event['kind']=='session' ? client.recordSession(payload) : client.recordTrace(payload); }
  Future<void> _append(Map<String, Object?> event) async { await config.storageDirectory.create(recursive:true); final rows=await _buffer.exists()?await _buffer.readAsLines():<String>[];rows.add(jsonEncode(event));await _buffer.writeAsString('${rows.reversed.take(100).toList().reversed.join('\n')}\n',flush:true); }
  Future<void> _replace(List<Map<String,dynamic>> events) async { await config.storageDirectory.create(recursive:true); final temp=File('${_buffer.path}.tmp');await temp.writeAsString(events.map(jsonEncode).join('\n'),flush:true);await temp.rename(_buffer.path); }

  @override void didChangeAppLifecycleState(AppLifecycleState state) { if (state==AppLifecycleState.resumed) { _session('foreground'); flush(); } else if (state==AppLifecycleState.paused || state==AppLifecycleState.detached) { _session('background',endedReason:state.name); } }
}
